source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc

```diff
@@ -148,10 +147,12 @@ std::chrono::microseconds GradientController::processLatencySamplesAndClear() {
 }
 
 uint32_t GradientController::calculateNewLimit() {
-  // Calculate the gradient value, ensuring it remains below the configured maximum.
   ASSERT(sample_rtt_.count() > 0);
-  const double raw_gradient = static_cast<double>(min_rtt_.count()) / sample_rtt_.count();
-  const double gradient = std::min(config_.maxGradient(), raw_gradient);
+
+  // Calculate the gradient value, ensuring it remains below the configured maximum.
+  const auto buffered_min_rtt = min_rtt_.count() + min_rtt_.count() * config_.minRTTBufferPercent();
+  const double raw_gradient = static_cast<double>(buffered_min_rtt) / sample_rtt_.count();
+  const double gradient = std::max<double>(0.5, std::min<double>(2.0, raw_gradient));
```

Comment:
How did you determine 0.5 as the lower bound for the clamp? Can you discuss this clamping of the gradient in the docs a bit? (Also per our offline convo I really do think a simulation utility would be very useful for evaluating algorithm changes)

Reply:
I put some thought into the simulation and keep coming back to something that looks like the bufferbloater tests, but implemented in the Envoy codebase to test the concurrency controllers. What complicates things here compared to the load balancer simulation is that the LB simulation is validating a certain request distribution after many requests, but an adaptive concurrency simulation requires us to validate its temporal behavior and how that would affect upstream latency.

This is achievable and I'll chip away at it here in the background. The next algorithm change will be prefaced with a simulator MVP.

Reply:
> Can you discuss this clamping of the gradient in the docs a bit?

Ping on this but feel free to do it as part of your next change.